File: tutorials/021_serve_genai_models/serve_genai_models.cpp
#include "neat/genai.h"

#include <cstdint>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

namespace genai = simaai::neat::genai;

struct Args {
  std::string host = "0.0.0.0";
  std::uint16_t port = 9998;
  std::filesystem::path llm;
  std::filesystem::path vlm;
  std::filesystem::path asr;
};

Args parse_args(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];
    if (arg == "--host" && i + 1 < argc) {
      args.host = argv[++i];
    } else if (arg == "--port" && i + 1 < argc) {
      args.port = static_cast<std::uint16_t>(std::stoi(argv[++i]));
    } else if (arg == "--llm" && i + 1 < argc) {
      args.llm = argv[++i];
    } else if (arg == "--vlm" && i + 1 < argc) {
      args.vlm = argv[++i];
    } else if (arg == "--asr" && i + 1 < argc) {
      args.asr = argv[++i];
    } else {
      throw std::runtime_error("usage: serve_genai_models [--host <host>] [--port <port>] "
                               "[--llm <dir>] [--vlm <dir>] [--asr <dir>]");
    }
  }
  if (args.llm.empty() && args.vlm.empty() && args.asr.empty()) {
    throw std::runtime_error("provide at least one of --llm <dir>, --vlm <dir>, or --asr <dir>");
  }
  return args;
}

int main(int argc, char** argv) {
  try {
    const Args args = parse_args(argc, argv);

    // STEP configure-server
    genai::GenAIServerOptions options;
    options.host = args.host;
    options.port = args.port;
    genai::GenAIServer server(options);
    // END STEP

    // STEP register-models
    if (!args.llm.empty()) {
      server.add_model(args.llm, "llm");
    }
    if (!args.vlm.empty()) {
      server.add_model(args.vlm, "vlm");
    }
    if (!args.asr.empty()) {
      server.add_model(args.asr, "asr");
    }

    std::cout << "registered models:";
    for (const auto& name : server.model_names()) {
      std::cout << " " << name;
    }
    std::cout << "\n";
    // END STEP

    // STEP start-serving
    std::cout << "serving on http://" << options.host << ":" << options.port << "\n";
    std::cout << "try: curl http://<modalix-ip>:" << options.port << "/v1/models\n";
    server.serve();
    // END STEP

    return 0;
  } catch (const std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/021_serve_genai_models/serve_genai_models.py
#!/usr/bin/env python3
import argparse
import time

import pyneat as neat


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", default="0.0.0.0")
    parser.add_argument("--port", type=int, default=9998)
    parser.add_argument("--llm", help="Path to an LLM model directory")
    parser.add_argument("--vlm", help="Path to a VLM model directory")
    parser.add_argument("--asr", help="Path to an ASR model directory")
    args = parser.parse_args()

    if not any([args.llm, args.vlm, args.asr]):
        raise RuntimeError("provide at least one of --llm, --vlm, or --asr")

    # STEP configure-server
    options = neat.genai.GenAIServerOptions()
    options.host = args.host
    options.port = args.port
    server = neat.genai.GenAIServer(options)
    # END STEP

    # STEP register-models
    if args.llm:
        server.add_model(args.llm, "llm")
    if args.vlm:
        server.add_model(args.vlm, "vlm")
    if args.asr:
        server.add_model(args.asr, "asr")

    print("registered models:", ", ".join(server.model_names()))
    # END STEP

    # STEP start-serving
    print(f"serving on http://{options.host}:{options.port}")
    print(f"try: curl http://<modalix-ip>:{options.port}/v1/models")
    server.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        server.stop()
    # END STEP

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
